### ISSUE TITLE
Slow performance compared to XMR-Stak-NVIDIA

### ISSUE THREAD

**Issue by stna1981** [fireice-uk/xmr-stak#748]
Hi,

I wanted to use XMR-Stak to mine to Nicehash, but the performance figures are really strange. I have 1 GTX1060 with 6GB and 4 GTX1060 with 3GB. They are all clocked at 1711 MHz core and +700 MHz memory, so the performance shpuöd be similar. The 6GB is set to 32 threads and 30 blocks, the 3GB cards at 32 threads and 27 blocks (as only 9 SMX).

Results with XMR-Stak:
0: 426  /  1: 529  /  2: 495  /  3: 400  /  4: 440    --> 2290 H/s

Results with XMR-Stak-NVIDIA:
0: 534  /  1: 526  /  2: 529  /  3: 531  /  4: 531    --> 2652 H/s

That's almost 14% more with the NVIDIA version of the miner. It's also strange that the cards yield so different in XMR-Stak. Any idea from dev site?

Best regards

Stefan

**Comment by psychocrypt** [comment]
Do you use the some config parameters for both miner versions?

**Comment by stna1981** [comment]
Regarding the GPU settings yes, both use the following conf:

"gpu_threads_conf" :
[
  { "index" : 0,
    "threads" : 32, "blocks" : 30,
    "bfactor" : 6, "bsleep" :  25,
    "affine_to_cpu" : false, "sync_mode" : 3,
  },
  { "index" : 1,
    "threads" : 32, "blocks" : 27,
    "bfactor" : 6, "bsleep" :  25,
    "affine_to_cpu" : false, "sync_mode" : 3,
  },
  { "index" : 2,
    "threads" : 32, "blocks" : 27,
    "bfactor" : 6, "bsleep" :  25,
    "affine_to_cpu" : false, "sync_mode" : 3,
  },
  { "index" : 3,
    "threads" : 32, "blocks" : 27,
    "bfactor" : 6, "bsleep" :  25,
    "affine_to_cpu" : false, "sync_mode" : 3,
  },
  { "index" : 4,
    "threads" : 32, "blocks" : 27,
    "bfactor" : 6, "bsleep" :  25,
    "affine_to_cpu" : false, "sync_mode" : 3,
  },
],

Only difference is the sync_mode property which is not available in XMR-Stak-NVIDIA.

**Comment by psychocrypt** [comment]
Sync mode 3 is equal to the hard coded sync mode in the old miner. We
changed a few kernel to solve memory issues maybe this hase an effect on
some gpus. I will check it soon

**Comment by stna1981** [comment]
Any update on this? :)

**Comment by psychocrypt** [comment]
curretly not. I am very bussy with other developments. I will check it
hopefully soon.

**Comment by stna1981** [comment]
As I will not be able to use XMR-Stak-NVIDIA any longer with the algo changes coming up, this issue really matters for me. Is there any update on this issue in the last release?

**Comment by Sally-Dawn-JR** [comment]
Hi there, same thing with 4 / GTX1070 8G. 

> Updated post fork and performance took a huge hit, roughly 16-17%. 

> All gpu settings carried over from XMR-RIG 1.x (NVIDIA OPTIMISED). 

Should we be hoping for a fix near future? Regards

**Comment by stna1981** [comment]
I dont think this is a bug, the performance decrease is due to the new algo. It's not for nothing that it's called CN "Heavy" ;-)

**Comment by stna1981** [comment]
@psychocrypt:
One assumption in #1359 was that the difference could come from the switch from CUDA 8 to CUDA 9 as the latest CUDA version focuses less on blockchain but more on AI, machine learning etc.

Is it the case that XMR-Stak-NVIDIA was using CUDA 8 whereas XMRStak is using CUDA 9? If yes, would it be possible to provide a CUDA 8 build? XMRStak-NVIDIA seems to have no issue with 8 cards and bFactor 7 while performance in XMRStak decreases by 5%...

**Comment by psychocrypt** [comment]
The release binaries are compiled with CUDA8. But if you think that CUDA8
is faster you should also switch the driver to CUDA8.
I don't think that it is faster but I am not sure.
We are mining now cryptinight_v7 it can be that you lose a few percent but
it depends on the gpu. I see no difference under linux on different GPUs.

**Comment by stna1981** [comment]
@psychocrypt: The problem was not that the hashrate is generally lower with XMRStak, but that - in my case - it is when 8 GPUs are used. Each GPU reaches 530 H/s when used alone, but with all 8, hashrate drops to about 500 for some of them. You said this might come from driver because of bFactor, but same driver and bFactor in XMRStak-NVIDIA do not lead to this issue. So CUDA8 was one guess what could be the reason. Maybe you have any better idea? :)

**Comment by psychocrypt** [comment]
No I have idea, I also saw this issue with muti nvidia gpus. Maybe it helps
to start two miner instances with each 4 gpus. But could be that the driver
is still the issue.

**Comment by stna1981** [comment]
I made a new comparison with 8x GTX 1060 3GB, all running at 1709/4600, with driver 391.35.

**XMRStak-NVIDIA:**
| ID |  10s |  60s |  15m | ID |  10s |  60s |  15m |
|  0 | 540.7 | 540.4 | 540.6 |  1 | 540.7 | 540.8 | 541.4 |
|  2 | 540.8 | 540.6 | 540.7 |  3 | 541.8 | 541.6 | 542.0 |
|  4 | 542.4 | 542.0 | 542.4 |  5 | 542.0 | 541.8 | 542.0 |
|  6 | 540.5 | 540.0 | 540.5 |  7 | 541.1 | 541.7 | 541.9 |
Totals:   4330.1 4328.8 4331.4 H/s

As you can see the hashrate is between 540.5 and 542.4, i. e. more or less identical.

**XMR-Stak 2.4.1:**
| ID |    10s |    60s |    15m | ID |    10s |    60s |    15m |
|  0 |  541.2 |  541.1 |  541.0 |  1 |  534.3 |  533.5 |  534.5 |
|  2 |  536.9 |  536.8 |  534.9 |  3 |  525.4 |  523.3 |  522.3 |
|  4 |  523.0 |  523.5 |  521.0 |  5 |  522.2 |  519.9 |  520.6 |
|  6 |  519.3 |  519.1 |  518.5 |  7 |  538.8 |  538.3 |  537.9 |
Totals (NVIDIA):  4241.0 4235.6 4230.6 H/s

Here the hashrate is between 519.3 and 541.2, i.e. a variace of 22 H/s or about 4%.
Still don't understand how the driver can be the reason if it performs nice with the older version...

**Comment by stna1981** [comment]
@psychocrypt:
Any comment on my observation (see post above in the EDIT)? :)

**Comment by Sally-Dawn-JR** [comment]
I can now verify with latest KlausT Nvidia CCminer (monero7 update) I got back to old hashrates. Also XMR-STAK 2.X crashes constantly (after having run old version quite successfully for months). Over and out!

**Comment by stna1981** [comment]
Any update?

**Comment by psychocrypt** [comment]
Do you tried the latest version 2.4.3 or better the dev branch? I can not
see any performance issue on my system.

**Comment by stna1981** [comment]
I can only use releases as I'm on windows and have no idea how to do any builds ;-) So yes I always use the latest version and also latest nvidia drivers. I still have the same issue that the hashrate is extremely different from card to card although they have the same settings and perform all better when running alone in the system. Example for CN-Heavy (RYO):

| ID |    10s |    60s |    15m | ID |    10s |    60s |    15m |
|  0 |  461.7 |  **460.7** |   (na) |  1 |  **439.1** |  440.0 |   (na) |
|  2 |  392.5 |  **393.0** |   (na) |  3 |  **392.1** |  381.7 |   (na) |
|  4 |  382.3 |  **361.6** |   (na) |  5 |  **398.9** |  402.8 |   (na) |
|  6 |  408.7 |  **400.7** |   (na) |  7 |  **441.2** |  441.9 |   (na) |
Totals (NVIDIA):  3316.6 3282.3    0.0 H/s

Worst card: 361 kH/s, best card: 460 kH/s -> a difference of almost 30%. When I test the cards solo, each card has about 460 kH/s, so I lose about 10% performance. CPU is at 20%, so this cannot be a bottleneck. RAM is 8GB, should also be fine. As I have 2 rigs with the same behaviour, I do not think it is a system issue.

But again: I did not have these issues with XMR-Stak-NVIDIA, that's why I believe it is a regression issue in XMR-Stak. Unfortunately XMR-Stak-NVIDIA is useless no since all the forking and ASICS. And also your XmrMiner does not have this issue. It is only XMRStak which has these performance issues with multiple cards, so maybe you can check again the differences between XmrMiner / XMR-Stak-NVIDIA and the newer XMR-Stak?

**Comment by psychocrypt** [comment]
I will have again a look into this issue. I also see this behaviour but do
not know where it is coming from.
